File: lib/data/local/daos/bible_dao.dart
import 'package:froom/froom.dart';

import '../entities/bible_entity.dart';

@dao
abstract class BibleDao {
  @Query('SELECT * FROM bibles')
  Future<List<BibleEntity>> getAllBibles();

  @Query('SELECT * FROM bibles WHERE isDownloaded = 1')
  Future<List<BibleEntity>> getDownloadedBibles();

  @Query('SELECT * FROM bibles WHERE isSelected = 1')
  Future<List<BibleEntity>> getSelectedBibles();

  @Query('SELECT * FROM bibles WHERE id = :id')
  Future<BibleEntity?> getBibleById(String id);

  @Insert(onConflict: OnConflictStrategy.replace)
  Future<void> insertBible(BibleEntity bible);

  @Insert(onConflict: OnConflictStrategy.replace)
  Future<void> insertBibles(List<BibleEntity> bibles);

  @Query('UPDATE bibles SET isDownloaded = 1 WHERE id = :id')
  Future<void> markAsDownloaded(String id);

  @Query('UPDATE bibles SET isSelected = 1 WHERE id = :id')
  Future<void> markAsSelected(String id);

  @Query('UPDATE bibles SET isSelected = 0 WHERE id = :id')
  Future<void> markAsUnselected(String id);

  @Query('DELETE FROM bibles WHERE id = :id')
  Future<void> deleteBible(String id);
}

File: core/database/src/main/java/com/biblelib/core/database/daos/BibleDao.kt
package com.biblelib.core.database.daos

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.biblelib.core.database.model.BibleEntity

@Dao
interface BibleDao {
    @Query("SELECT * FROM bibles ORDER BY sortOrder ASC")
    suspend fun getAll(): List<BibleEntity>

    @Query("SELECT * FROM bibles WHERE abbreviation = :abbr LIMIT 1")
    suspend fun getByAbbr(abbr: String): BibleEntity?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(bible: BibleEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(records: List<BibleEntity>)

    @Delete
    suspend fun delete(bible: BibleEntity)

    @Query("DELETE FROM bibles WHERE abbreviation = :abbr")
    suspend fun deleteByAbbr(abbr: String)

    @Query("DELETE FROM bibles")
    suspend fun deleteAll()

    @Query("UPDATE bibles SET isDownloaded = 1, downloadFailed = 0, downloadProgress = 1.0 WHERE abbreviation = :abbr")
    suspend fun markDownloaded(abbr: String)

    @Query("UPDATE bibles SET downloadProgress = :progress, downloadFailed = 0 WHERE abbreviation = :abbr")
    suspend fun updateProgress(abbr: String, progress: Float)

    @Query("UPDATE bibles SET downloadFailed = 1, downloadProgress = :progress WHERE abbreviation = :abbr")
    suspend fun markFailed(abbr: String, progress: Float)

    @Query("UPDATE bibles SET downloadFailed = 0 WHERE abbreviation = :abbr")
    suspend fun clearFailed(abbr: String)
}